### ISSUE TITLE
no alloc support

### ISSUE THREAD

**Issue by theunkn0wn1** [capnproto/capnproto-rust#221]
I note #71 Which adds support for `no_std` crates, but `capnp` still depends on an allocator.
It is desirable to have support for environments that do not have an allocator / are memory constrained such as microcontrollers. 

In my application, I am presently using `postcard` to provide object de/encoding on both sides of the wire. Among other features, it supports de/encoding messages out of a `&[u8]` buffer, thus allowing it to work without alloc. 

I am evaluating capnproto to see if it can be a viable replacement for `postcard` and gRPC (replacing the latter with capnproto's RPC) as to simplify my software situation.

It appears that `capnp`, even in `no_std` mode, does not have a mechanism for statically parsing messages.

**Comment by dwrensha** [comment]
Yes, that is accurate: the `capnp` crate still requires an allocator. I agree that it would be good to remove that requirement. To achieve that, we would need:
 1. An implementation of [`ReaderSegments`](https://github.com/capnproto/capnproto-rust/blob/eec64b6e55543fa4e6bdb047c49df957f764bdc3/capnp/src/message.rs#L93) that does not allocate
 2.  To remove the [`String`](https://github.com/capnproto/capnproto-rust/blob/eec64b6e55543fa4e6bdb047c49df957f764bdc3/capnp/src/lib.rs#L161) from `capnp::Error`.

Note, however, that getting the `capnp-rpc` crate (as opposed to just the base crate `capnp`) to work with `no_std` and no `alloc` would be a lot more difficult.

**Comment by theunkn0wn1** [comment]
Thanks for the response.

For Capnproto to be viable, I don't require RPC support in no std; I only require the capability to statically decode a message of known maximum size.

**Comment by romixlab** [comment]
Tried it on STM32F051 (thumbv6m) with alloc-cortex-m crate. Unfortunately something in `capnp::private::layout::wire_helpers::allocate ()` causes a HardFault to occur on `str	r1, [r5, #0]` instruction, while r1 = 0.

Btw size of the binary increased by approx 10kB after adding allocator and capnp dependencies (with opt-level = "z", codegen-units = 1, lto = true)

**Comment by dwrensha** [comment]
We got another report of a hard fault using the alloc-cortex-m crate here: #293

**Comment by dwrensha** [comment]
@romixlab I expect there's a good chance that if you upgrade to capnp-v0.14.10 the HardFault will go away (and probably be replaced by an allocation failure).

**Comment by romixlab** [comment]
Would be interesting to check again!

**Comment by ariel-miculas** [comment]
> 2. To remove the [`description: String`](https://github.com/capnproto/capnproto-rust/blob/eec64b6e55543fa4e6bdb047c49df957f764bdc3/capnp/src/lib.rs#L161) field from `capnp::Error`, maybe replacing it with an `enum`.

@dwrensha I'm trying to integrate capnproto into the linux kernel, where there is an alloc crate but with limited functionality (e.g. all the allocations are fallible, so try_new is used instead of new; there is also no `String`). I was thinking about the same thing, using an `enum` instead of descriptions for errors; are you willing to accept such a PR (in order to minimize the differences between the kernel version and the upstream version). Note that it's not yet clear whether capnproto could be accepted in the kernel, for the moment I'm mostly experimenting.

**Comment by dwrensha** [comment]
Yes, I still want this to happen, and I would review a PR.

A question that I had been stuck on is: if `capnp::Error` is no-alloc, then what do we do about the `capnp_rpc::rpc::remote_exception_to_error()` function?
https://github.com/capnproto/capnproto-rust/blob/3bbb520205a661987e23b6d06d2b0c646b62f6c5/capnp-rpc/src/rpc.rs#L368-L384

The `reason` field is an arbitrary `Text` from a capnp message, and it would be sad to lose any information contained in it.

Just now I had the thought: `capnp::Error` could accommodate this with a variant that only exists when the `alloc` feature is enabled.

**Comment by dwrensha** [comment]
Hm... that would work, but I'm worried that it would end up with a lot of `#[cfg]`-gated code. I was imagining instead that the `String` would be part of the enum variant. Like
```
pub enum Error {
  IncompatibileTypes,
  OutOfBounds,
  // ...
  #[cfg(feature="alloc")]
  RpcError(String),
}
```

The `Debug` implementation for `capnp::Error` can expand the variants into strings like "Existing list value is compatible...". Moreover, the variants can have nontrivial data, like integer offsets or type IDs, which can be incorporated into the output of the `Debug` impl.

**Comment by ariel-miculas** [comment]
Yes, but what I'm trying to do is to avoid allocating a `String` for the error, that's why I wanted to replace the error handling with an enum (which could still hold data, just not a `String`).

**Comment by dwrensha** [comment]
This is fixed in #417. We are aiming to release v0.18 with this change in the next week or so.

**Comment by dwrensha** [comment]
no-alloc support was released just now in version 0.18.0
